src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -290,8 +297,120 @@ OCC::Optional<OCC::PinStateEnums::PinState> OCC::CfApiWrapper::PlaceHolderInfo::
+
+    const QString providerSyncRootIdRegistryKey = QString() % R"(SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\SyncRootManager\)" % syncRootId;
+    const QString providerSyncRootIdUserSyncRootsRegistryKey = QString() % providerSyncRootIdRegistryKey % R"(\UserSyncRoots\)";
+
+    const QVector<std::tuple<HKEY, QString, QString, int, QVariant>> registryKeysToSet = {
+        { HKEY_LOCAL_MACHINE, providerSyncRootIdRegistryKey, QStringLiteral("Flags"), REG_DWORD, 34 },
+        { HKEY_LOCAL_MACHINE, providerSyncRootIdRegistryKey, QStringLiteral("DisplayNameResource"), REG_EXPAND_SZ, displayName },
+        { HKEY_LOCAL_MACHINE, providerSyncRootIdRegistryKey, QStringLiteral("IconResource"), REG_EXPAND_SZ, QString(QDir::toNativeSeparators(qApp->applicationFilePath()) + QString(",0")) },
+        { HKEY_LOCAL_MACHINE, providerSyncRootIdUserSyncRootsRegistryKey, windowsSID, REG_SZ, syncRootPath }
+    };
+
+    bool isAnyKeySetFailed = false;
+
+    for (const auto &registryKeyToSet : qAsConst(registryKeysToSet)) {
+        HKEY rootKey;
+        QString subKey;
+        QString valueName;
+        int type;
+        QVariant value;
+
+        std::tie(rootKey, subKey, valueName, type, value) = registryKeyToSet;
+
+        if ((isAnyKeySetFailed = !OCC::Utility::registrySetKeyValue(rootKey, subKey, valueName, type, value))) {
+            break;
+        }
+    }
+
+    if (isAnyKeySetFailed) {
+        qCWarning(lcCfApiWrapper) << "Failed to set Registry keys for shell integration. Progress bar will not work.";
+        Q_ASSERT(!OCC::Utility::registryDeleteKeyTree(HKEY_LOCAL_MACHINE, providerSyncRootIdRegistryKey));
+    } else {
+        registeredSyncRootKeys[syncRootPath] = providerSyncRootIdRegistryKey;
+        qCInfo(lcCfApiWrapper) << "Successfully set Registry keys for shell integration at:" << providerSyncRootIdRegistryKey << ". Progress bar will work.";
+    }
+
+    return !isAnyKeySetFailed;
+}
+
+bool deleteSyncRootRegistryKey(const QString& syncRootPath)
+{
+    const auto foundRegisteredSyncRootKey = registeredSyncRootKeys.find(syncRootPath);
+    if (foundRegisteredSyncRootKey != registeredSyncRootKeys.end() && !foundRegisteredSyncRootKey->isEmpty()) {
+        bool result = OCC::Utility::registryDeleteKeyTree(HKEY_LOCAL_MACHINE, *foundRegisteredSyncRootKey);
+        registeredSyncRootKeys.remove(*foundRegisteredSyncRootKey);
+        return result;
+    }
+
+    return true;
+}
+
+OCC::Result<void, QString> OCC::CfApiWrapper::registerSyncRoot(const QString &path, const QString &providerName, const QString &providerVersion, const QString &folderAlias, const QString &displayName, const QString &accountDisplayName)
+{
+    // even if we fail to register our sync root with shell, we can still proceed with using the VFS
+    Q_ASSERT(createSyncRootRegistryKeys(providerName, folderAlias, displayName, accountDisplayName, path));
```

Comment:
Dangerous! This won't get executed in release builds.

Reply:
My point was that: `createSyncRootRegistryKeys` won't even get called in release build. It's not just about asserting or not, the whole expression is compiled out.

Reply:
> My point was that: `createSyncRootRegistryKeys` won't even get called in release build. It's not just about asserting or not, the whole expression is compiled out.

@er-vin Indeed. Fixed.